db/migrate/20210201142920_add_public_id_to_qualifications.rb

```diff
@@ -0,0 +1,6 @@
+class AddPublicIdToQualifications < ActiveRecord::Migration[6.0]
+  def change
+    add_column :application_qualifications, :constituent_grades, :jsonb
+    add_column :application_qualifications, :public_id, :bigint
```

Comment:
Following on from our chat, if you'd like to use postgres sequences to ensure you always have unique ids for each qualification, you could define a sequence-based default here:

```suggestion
    add_column :application_qualifications, :public_id, :bigint, default: -> { "nextval('qualifications_public_id_seq')" }
```

You may have to define this sequence first in your db migration, along the lines of:
```
  execute "CREATE SEQUENCE qualifications_public_id_seq;"
```
and you could also set its start to take into account any existing `public_id`s, if any:

```
  execute "SELECT setval('qualifications_public_id_seq', max(public_id)) FROM application_qualifications;"
```
This should go at the end of the migration. Setting a default will also populate all `public_id`s with unique ascending numbers at the point of migration. If you'd like to use different ranges for `public_id` there are ways to achieve this (e.g. you could set a starting point for the sequence with START when defining it).

For setting nested `public_id`s in the jsonb column post-migration, you'll need to be calling ActiveRecord::Base.connection.execute("SELECT nextval('qualifications_public_id_seq')") in the ruby code for each new public_id you set. But this way you get unique `public_id` numbers for both the top-level column and the nested `public_id` values.

Hope this helps!

Reply:
Thanks, this is really useful and will make the public_id setting way more performant!

- We don't want to use the default here, because we don't necessarily want to set the value for public_id in the column (for structured qualifications it will be nil, and set inside the JSON instead)
- I'm going to add a service to handle setting the public_ids. The initial migration will call this on existing qualifications, and from then on, when a qualification is created it will be called. I can use the sequence there – it will just be less baked-in to the DB